src/clients/node/src/node.zig

```diff
@@ -15,783 +19,472 @@ const CreateTransfersResult = tb.CreateTransfersResult;
-
-    return null;
-}
-
-fn on_result(user_data: u128, operation: Operation, results: []const u8) void {
-    // A reference to the user's JS callback was made in `request` or `raw_request`. This MUST be
-    // cleaned up regardless of the result of this function.
-    const env = @as(translate.UserData, @bitCast(user_data)).env;
-    const callback_reference = @as(translate.UserData, @bitCast(user_data)).callback_reference;
-    defer translate.delete_reference(env, callback_reference) catch {
-        std.log.warn("on_result: Failed to delete reference to user's JS callback.", .{});
-    };
 
-    const napi_callback = translate.reference_value(
-        env,
-        callback_reference,
-        "Failed to get callback reference.",
-    ) catch return;
-    const scope = translate.scope(
-        env,
-        "Failed to get \"this\" for results callback.",
-    ) catch return;
-    const globals_raw = translate.globals(env) catch return;
-    const globals = globalsCast(globals_raw.?);
-    const argc: usize = 2;
-    var argv: [argc]c.napi_value = undefined;
-
-    const napi_results = switch (operation) {
-        .create_accounts => encode_napi_results_array(
-            CreateAccountsResult,
-            env,
-            results,
-        ) catch return,
-        .create_transfers => encode_napi_results_array(
-            CreateTransfersResult,
-            env,
-            results,
-        ) catch return,
-        .lookup_accounts => encode_napi_results_array(Account, env, results) catch return,
-        .lookup_transfers => encode_napi_results_array(Transfer, env, results) catch return,
-    };
+    // Allocate enough memory to contain the Events and the Results.
+    const alloc_size = @max(@sizeOf(Event) * array_length, @sizeOf(Result) * array_length);
+    if (@sizeOf(vsr.Header) + alloc_size > constants.message_size_max) {
+        return translate.throw(env, "Batch is larger than the maximum message size.");
+    }
 
-    argv[0] = globals.napi_undefined;
-    argv[1] = napi_results;
+    const alloc_align = @max(@alignOf(Event), @alignOf(Result));
+    comptime assert(alloc_align <= 16); // Assume 16-byte alignment. Makes free() simpler.
 
-    translate.call_function(env, scope, napi_callback, argc, argv[0..]) catch {
-        translate.throw(env, "Failed to call JS results callback.") catch return;
+    const data = allocator.alignedAlloc(u8, 16, alloc_size) catch |err| switch (err) {
+        error.OutOfMemory => return translate.throw(env, "Batch allocation ran out of memory."),
     };
-}
-
-fn tick(env: c.napi_env, info: c.napi_callback_info) callconv(.C) c.napi_value {
-    var argc: usize = 1;
-    var argv: [1]c.napi_value = undefined;
-    if (c.napi_get_cb_info(env, info, &argc, &argv, null, null) != c.napi_ok) {
-        translate.throw(env, "Failed to get args.") catch return null;
+    errdefer allocator.free(data);
```

Comment:
Suggestion: rename `data` to `buffer` and use a block to keep the related code/comments together.
I think `data` is ambiguous with `events`.

```zig
    const buffer: []align(16) u8 = buffer: {
        // Allocate enough memory to contain the Events and the Results.
        const alloc_size = @max(@sizeOf(Event) * array_length, @sizeOf(Result) * array_length);
        if (@sizeOf(vsr.Header) + alloc_size > constants.message_size_max) {
            return translate.throw(env, "Batch is larger than the maximum message size.");
        }

        const alloc_align = @max(@alignOf(Event), @alignOf(Result));
        comptime assert(alloc_align <= 16); // Assume 16-byte alignment. Makes free() simpler.

        break :buffer allocator.alignedAlloc(u8, 16, alloc_size) catch |err| switch (err) {
            error.OutOfMemory => return translate.throw(env, "Batch allocation ran out of memory."),
        };
    };
    errdefer allocator.free(buffer);
```